crypto/lhash/lhash.c

```diff
@@ -186,16 +186,35 @@ void OPENSSL_LH_doall_arg(OPENSSL_LHASH *lh, OPENSSL_LH_DOALL_FUNCARG func, void
 static int expand(OPENSSL_LHASH *lh)
 {
     OPENSSL_LH_NODE **n, **n1, **n2, *np;
-    unsigned int p, i, j;
-    unsigned long hash, nni;
+    unsigned int p, pmax, nni, i, j;
+    unsigned long hash;
+
+    nni = lh->num_alloc_nodes;
+    p = lh->p;
+    pmax = lh->pmax;
+    if (p + 1 >= pmax) {
+        j = nni * 2;
+        n = OPENSSL_realloc(lh->b, sizeof(OPENSSL_LH_NODE *) * j);
+        if (n == NULL) {
+            lh->error++;
+            return 0;
+        }
+        for (i = nni; i < j; i++)
+            n[i] = NULL;
+        lh->pmax = lh->num_alloc_nodes;
+        lh->num_alloc_nodes = j;
+        lh->num_expand_reallocs++;
+        lh->p = 0;
+        lh->b = n;
+    } else {
+        lh->p++;
```

Comment:
I am rather perplexed as to why we set `lh->p = 0` when reallocating, and increment it instead otherwise?  Should it not end up at the same value either way?  Why should the resulting state be different if we needed a realloc than when we did not?  What is `lh->p` for?

Reply:
It's a great question. After over an hour of studying this code I still could not figure out exactly what `p` is used for or why it works the way it does. It appears to be some kind of counter that increments each time we add nodes. As soon as `p` hits `pmax` we add more nodes, update `pmax` and reset the counter `p`. That's why things are different in the realloc case, i.e. we are resetting the counter, whereas otherwise we just increment it.

In the end I opted for maintaining the current behaviour (this is actually what the current code does - it increments `p` unconditionally, and then resets it to 0 in the event of a realloc). This seemed the simplest solution.

Reply:
OK, it seems that `p` is the first index for which the hashes may need to be recomputed after the hash table expands.  If a number is congruent to `x mod n` then it is either `x` or `x + n` mod `2n` so the hash table update moves some nodes from `x` to `x + n` where `n` is `pmax` and the table size is `2 * pmax`.

There are way too many redundant poorly documented variables here.  This code needs to be put on the urgent need for repair list.

Reply:
Hash nodes with  `hash < p mod pmax` have up-to-date hash slots based on the most recent expansion of the table.  The others are hashed with the previous modulus.  As we call expand repeatedly, eventually `p == pmax` and all the hashes become current, at which point we actually grow the array and invalidate all the hashes relative to the new modulus (so they all continue to use the old modulus).